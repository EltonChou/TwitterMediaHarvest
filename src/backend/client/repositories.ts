import { Sha256 } from '@aws-crypto/sha256-browser'
import type { CognitoIdentityCredentials } from '@aws-sdk/credential-provider-cognito-identity'
import { FetchHttpHandler, streamCollector } from '@aws-sdk/fetch-http-handler'
import { HttpRequest } from '@aws-sdk/protocol-http'
import { SignatureV4 } from '@aws-sdk/signature-v4'
import { CreateClientFailed, UpdateStatsFailed } from '@backend/errors'
import type { IStorageProxy } from '@libs/proxy'
import type { ClientInfo, V4Statistics } from '@schema'
import jws from 'jws'
import type { Storage } from 'webextension-polyfill'
import Browser from 'webextension-polyfill'
import { ClientInfoVO } from './valueObjects'

type UpdateInfo = Pick<ClientInfo, 'csrfToken' | 'syncedAt'>

const defaultInfo: Readonly<ClientInfo> = {
  uuid: 'uuid',
  csrfToken: 'token',
  syncedAt: 0,
  uninstallCode: 'code',
}

const isEmptyInfo = (record: Record<string, unknown>) =>
  record['uuid'] === defaultInfo.uuid ||
  record['csrfToken'] === defaultInfo.csrfToken ||
  record['syncedAt'] === defaultInfo.syncedAt ||
  record['uninstall'] === defaultInfo.uninstallCode

type ProviderOptions = {
  credentialProvider: Provider<CognitoIdentityCredentials>
  statsProvider: Provider<V4Statistics>
}

export interface IClientInfoRepository {
  getInfo(options?: ProviderOptions): Promise<ClientInfoVO>
  updateStats(options?: ProviderOptions): Promise<void>
  resetInfo(): Promise<void>
}

export class InfoSyncLock implements IProcessLock {
  private lockCriteria = 'InfoUpdateLock'
  private maxLockTime = 10 * 60 * 1000

  constructor(readonly storageArea: Storage.StorageArea) {}

  async isLocked(): Promise<boolean> {
    const record = await this.storageArea.get(this.lockCriteria)
    return Object.keys(record).includes(this.lockCriteria) && record[this.lockCriteria] - Date.now() >= this.maxLockTime
  }

  async release(): Promise<void> {
    await this.storageArea.remove(this.lockCriteria)
  }

  async acquire(): Promise<void> {
    await this.storageArea.set({ [this.lockCriteria]: Date.now() })
  }
}

export class ClientInfoRepository implements IClientInfoRepository {
  constructor(readonly storageArea: IStorageProxy<ClientInfo>, private defaultOptions: ProviderOptions) {}

  private async createClient(options: ProviderOptions): Promise<ClientTokenResponse> {
    const initStats = await options.statsProvider()
    const credential = await options.credentialProvider()

    const handler = new ClientApiHandler()
    const response = await handler.send(HttpMethod.Post, '/clients', {}, initStats, credential)

    if (response.statusCode === 201) {
      const body = JSON.parse(response.body)
      return body
    }

    throw new CreateClientFailed(
      response.statusCode,
      response.body?.message || JSON.stringify(response.body),
      response.headers
    )
  }

  private async getNewInfo(options?: Partial<ProviderOptions>): Promise<ClientInfoVO> {
    const clientTokenResponse: ClientTokenResponse = await this.createClient({ ...this.defaultOptions, ...options })
    const { payload } = jws.decode(clientTokenResponse.token, { json: true })
    const info = new ClientInfoVO({
      uuid: payload['uuid'],
      csrfToken: clientTokenResponse.token,
      syncedAt: Date.now(),
      uninstallCode: clientTokenResponse?.uninstallCode || payload['uninstallCode'],
    })
    await this.storageArea.setItem(info.props)
    await Browser.runtime.setUninstallURL(info.uninstallUrl)
    return info
  }

  async getInfo(options?: Partial<ProviderOptions>): Promise<ClientInfoVO> {
    const record = await this.storageArea.getItemByDefaults(defaultInfo)
    return isEmptyInfo(record) ? this.getNewInfo(options) : new ClientInfoVO(record)
  }

  async updateStats(options?: Partial<ProviderOptions>): Promise<void> {
    options = { ...this.defaultOptions, ...options }

    const { uuid, csrfToken } = await this.getInfo(options)
    const stats = await options.statsProvider()
    const credential = await options.credentialProvider()

    const handler = new ClientApiHandler()
    const response = await handler.send(
      HttpMethod.Put,
      '/clients/' + uuid + '/stats',
      { 'x-csrf-token': csrfToken },
      stats,
      credential
    )

    if (response.statusCode !== 200)
      throw new UpdateStatsFailed(
        response.statusCode,
        response.body?.message || JSON.stringify(response.body),
        response.headers
      )

    const body = JSON.parse(response.body)
    const clientInfo: UpdateInfo = {
      csrfToken: body.token,
      syncedAt: Date.now(),
    }
    await this.storageArea.setItem(clientInfo)
  }

  async resetInfo(): Promise<void> {
    await this.storageArea.setItem(defaultInfo)
  }
}

enum HttpMethod {
  Post = 'POST',
  Get = 'GET',
  Put = 'PUT',
  Delete = 'DELETE',
}

class ClientApiHandler {
  async send(
    method: HttpMethod,
    path: string,
    headers: Record<string, string>,
    body: unknown,
    credentials: CognitoIdentityCredentials
  ) {
    const request = new HttpRequest({
      protocol: 'https',
      body: JSON.stringify(body),
      headers: {
        'Content-Type': 'application/json',
        'X-Api-Key': process.env.API_KEY,
        'X-MediaHarvest-Version': process.env.VERSION_NAME,
        host: process.env.API_HOSTNAME,
        ...headers,
      },
      method: method,
      hostname: process.env.API_HOSTNAME,
      path: (process.env.API_ROOT_PATH || '') + path,
    })

    const signer = new SignatureV4({
      region: 'ap-northeast-1',
      service: 'execute-api',
      credentials: credentials,
      sha256: Sha256,
    })

    const signedRequest = await signer.sign(request)

    const client = new FetchHttpHandler()
    const { response } = await client.handle(signedRequest as HttpRequest)

    response.body = new TextDecoder().decode(await streamCollector(response.body))
    return response
  }
}
