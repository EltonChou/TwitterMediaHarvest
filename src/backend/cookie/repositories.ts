import browser from 'webextension-polyfill'

export interface ITwitterTokenRepository {
  getCsrfToken(): Promise<string | null>
  getGuestToken(): Promise<string | null>
}

/**
 * @deprecated
 */
export class TwitterTokenRepository implements ITwitterTokenRepository {
  async getCsrfToken(): Promise<string | null> {
    const cookie = await browser.cookies.get({
      url: 'https://twitter.com',
      name: 'ct0',
    })

    return cookie ? cookie.value : null
  }

  async getGuestToken(): Promise<string | null> {
    const cookie = await browser.cookies.get({
      url: 'https://twitter.com',
      name: 'gt',
    })
    return cookie ? cookie.value : null
  }
}

export class XTokenRepository implements ITwitterTokenRepository {
  async getCsrfToken(): Promise<string | null> {
    const cookie = await browser.cookies.get({
      url: 'https://x.com',
      name: 'ct0',
    })

    return cookie ? cookie.value : null
  }

  async getGuestToken(): Promise<string | null> {
    const cookie = await browser.cookies.get({
      url: 'https://x.com',
      name: 'gt',
    })
    return cookie ? cookie.value : null
  }
}
