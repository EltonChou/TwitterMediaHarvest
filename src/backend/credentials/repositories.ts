import type { CognitoIdentityCredentials, Storage } from '@aws-sdk/credential-provider-cognito-identity'
import { fromCognitoIdentityPool } from '@aws-sdk/credential-providers'
import ValueObject from '@backend/valueObject'
import type { AWSCredentials } from '@schema'
import type { Storage as BrowserStorage } from 'webextension-polyfill'
import Browser from 'webextension-polyfill'

class CredentialVO extends ValueObject<AWSCredentials> {
  readonly expireTimeRedundancy: number = 3 * 60 * 1000

  constructor(props: AWSCredentials) {
    super(props)
  }

  get credential(): CognitoIdentityCredentials {
    return { ...this.props, expiration: new Date(this.props.expiration) }
  }

  get isExpired(): boolean {
    return this.props?.expiration ? this.props.expiration < Date.now() - this.expireTimeRedundancy : true
  }

  static fromCognitoIdentityCredentials(credential: CognitoIdentityCredentials): CredentialVO {
    return new CredentialVO({
      ...credential,
      // ? Does it means the credential wouldn't expired if the expiration was undefined?
      expiration: credential?.expiration ? credential.expiration.getTime() : 0,
    })
  }
}

export interface ICredentialRepository {
  getCredential(): Promise<CognitoIdentityCredentials>
}

class BrowserSyncStorageProxy implements Storage {
  async getItem(key: string): Promise<string | null> {
    const record = await Browser.storage.sync.get(key)
    return Object.keys(record).length === 0 ? null : record[key]
  }

  async removeItem(key: string): Promise<void> {
    return await Browser.storage.sync.remove(key)
  }

  async setItem(key: string, data: string): Promise<void> {
    return await Browser.storage.sync.set({ [key]: data })
  }
}

export class CredentialRepository implements ICredentialRepository {
  readonly credentialCriteria = 'awsCredential'

  constructor(readonly storageArea: BrowserStorage.StorageArea) {}

  private async fetchCredential(): Promise<CognitoIdentityCredentials> {
    // ! Data in storage is `AWSCredential` not `CognitoIdentityCredentials`
    // TODO: Need to define the schema of storage area.
    const credential = await fromCognitoIdentityPool({
      identityPoolId: process.env.IDENTITY_POOL_ID,
      cache: new BrowserSyncStorageProxy(),
      clientConfig: {
        region: process.env.IDENTITY_POOL_REGION,
      },
    })()

    const credentialVO = CredentialVO.fromCognitoIdentityCredentials(credential)
    await this.storageArea.set({ [this.credentialCriteria]: credentialVO.props })
    return credential
  }

  async getCredential(): Promise<CognitoIdentityCredentials> {
    const credentialRecord = await this.storageArea.get(this.credentialCriteria)
    if (Object.keys(credentialRecord).length === 0) return await this.fetchCredential()

    const credentialsVO = new CredentialVO(credentialRecord[this.credentialCriteria])
    return credentialsVO.isExpired ? await this.fetchCredential() : credentialsVO.credential
  }
}
