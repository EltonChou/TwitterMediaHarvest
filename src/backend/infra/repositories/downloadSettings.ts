import type { ISettingsRepository } from '@backend/domain/repositories/settings'
import type { IStorageProxy } from '@libs/proxy'
import type { DownloadSettings } from '@schema'

const defaultIntegrationSettings: DownloadSettings = {
  enableAria2: false,
  aggressiveMode: false,
  askWhereToSave: false,
}

export class DownloadSettingsRepository implements ISettingsRepository<DownloadSettings> {
  constructor(readonly storageArea: IStorageProxy<DownloadSettings>) {}

  async get(): Promise<DownloadSettings> {
    const settings = await this.storageArea.getItemByDefaults(defaultIntegrationSettings)
    return settings as DownloadSettings
  }

  async save(settings: Partial<DownloadSettings>): Promise<void> {
    await this.storageArea.setItem(settings)
  }

  async reset(): Promise<void> {
    await this.save(defaultIntegrationSettings)
  }

  getDefault(): DownloadSettings {
    return defaultIntegrationSettings
  }
}
