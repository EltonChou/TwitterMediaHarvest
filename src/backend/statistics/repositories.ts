import {
  BrowserStorageFetcher,
  BrowserStorageSetter,
  storageFetcher,
  storageSetter
} from '../../libs/chromeApi'

export enum StatisticsKey {
  SuccessDownloadCount = 'successDownloadCount',
  FailedDownloadCount = 'failedDownloadCount',
  ErrorCount = 'errorCount',
}

type DownloadStatistic = {
  [StatisticsKey.SuccessDownloadCount]?: number
  [StatisticsKey.FailedDownloadCount]?: number
  [StatisticsKey.ErrorCount]?: number
}

const defaultStatistic: DownloadStatistic = {
  [StatisticsKey.ErrorCount]: 0,
  [StatisticsKey.FailedDownloadCount]: 0,
  [StatisticsKey.SuccessDownloadCount]: 0,
}

export interface IStatisticsRepository {
  getStatisticsCount(key: StatisticsKey): Promise<number>
  addStatisticsCount(key: StatisticsKey): Promise<void>
}

export default class StatisticsRepository implements IStatisticsRepository {
  private fetchStorage: BrowserStorageFetcher
  private setStorage: BrowserStorageSetter

  constructor(storageArea: chrome.storage.StorageArea) {
    this.fetchStorage = storageFetcher(storageArea)
    this.setStorage = storageSetter(storageArea)
  }

  async getStatisticsCount(
    key: StatisticsKey
  ): Promise<number> {
    const downloadStatistic: DownloadStatistic = {}
    downloadStatistic[key] = 0

    const count = await this.fetchStorage(downloadStatistic)
    return count[key]
  }

  async addStatisticsCount(key: StatisticsKey): Promise<void> {
    const count = await this.getStatisticsCount(key)
    const downloadCount: DownloadStatistic = {}
    downloadCount[key] = count + 1

    await this.setStorage(downloadCount)
  }

  async setDefaultStatistics(): Promise<void> {
    await this.setStorage(defaultStatistic)
  }
}