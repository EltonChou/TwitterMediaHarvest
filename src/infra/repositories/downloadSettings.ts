/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import type { ISettingsRepository } from '#domain/repositories/settings'
import type { IStorageProxy } from '#libs/storageProxy'
import type { DownloadSettings } from '#schema'

const defaultIntegrationSettings: DownloadSettings = {
  enableAria2: false,
  aggressiveMode: false,
  askWhereToSave: false,
}

export class DownloadSettingsRepository
  implements ISettingsRepository<DownloadSettings>
{
  constructor(readonly storageArea: IStorageProxy<DownloadSettings>) {}

  async get(): Promise<DownloadSettings> {
    const settings = await this.storageArea.getItemByDefaults(
      defaultIntegrationSettings
    )
    return settings
  }

  async save(settings: Partial<DownloadSettings>): Promise<void> {
    await this.storageArea.setItem(settings)
  }

  async reset(): Promise<void> {
    await this.save(defaultIntegrationSettings)
  }

  getDefault(): DownloadSettings {
    return defaultIntegrationSettings
  }
}
