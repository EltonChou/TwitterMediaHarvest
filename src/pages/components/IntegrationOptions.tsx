import type { ISettingsRepository } from '#domain/repositories/settings'
import type { IWarningSettingsRepo } from '#domain/repositories/warningSettings'
import { isFirefox } from '#helpers/runtime'
import { getText as i18n } from '#libs/i18n'
import useDownloadSettings from '#pages/hooks/useDownloadSettings'
import useWarningSettings from '#pages/hooks/useWarningSettings'
import Links from '#pages/links'
import { testIdProps } from '#pages/utils'
import type { DownloadSettings } from '#schema'
import type { HelperMessage } from './controls/featureControls'
import { RichFeatureSwitch } from './controls/featureControls'
import { Link, VStack } from '@chakra-ui/react'
import React from 'react'

const Aria2Description = () => {
  const aria2Ext = i18n('Aria2-Explorer', 'options:integrations')
  const full = i18n(
    'Transfer the download to Aria2 via {{aria2-extension}}.',
    'options:integrations',
    { 'aria2-extension': aria2Ext }
  )

  if (process.env.TARGET === 'firefox') return <>{full}</>

  const [prefix, suffix] = full.split(aria2Ext)
  return (
    <>
      {prefix}
      <Link
        href={Links.aria2Explorer}
        target="_blank"
        color={'blue.400'}
        {...testIdProps('aria2-ext-link')}
      >
        {aria2Ext}
      </Link>
      {suffix}
    </>
  )
}

type IntegrationOptionsProps = {
  downloadSettingsRepo: ISettingsRepository<DownloadSettings>
  warningSettingsRepo: IWarningSettingsRepo
}

/**
 * Some switches is disabled when `process.env.TARGET` is `firefox`.
 */
const IntegrationOptions = (props: IntegrationOptionsProps) => {
  const { settings: integrationSettings, toggler: downloadSettingsToggler } =
    useDownloadSettings(props.downloadSettingsRepo)
  const { settings: warningSettings, toggler: warningSettingsToggler } =
    useWarningSettings(props.warningSettingsRepo)
  const isInFireFox = isFirefox()

  const message: HelperMessage | undefined = isInFireFox
    ? {
        type: 'info',
        content: i18n(
          'This integration is not compatible with {{platform}}',
          'options:integrations',
          { platform: 'Firefox' }
        ),
      }
    : undefined

  return (
    <VStack>
      <RichFeatureSwitch
        name={i18n('Dispatch download to Aria2', 'options:integrations')}
        desc={<Aria2Description />}
        isOn={integrationSettings.enableAria2}
        handleClick={downloadSettingsToggler.enableAria2}
        isDisable={isInFireFox}
        message={message}
        testId="dispatchToAria2-integration-switch"
      />
      <RichFeatureSwitch
        name={i18n('Aggressive Mode', 'options:integrations')}
        desc={i18n(
          'Ensure the filename not to be modified by other extensions. This mode might be conflicted with other download management extensions.',
          'options:integrations'
        )}
        isOn={integrationSettings.aggressiveMode}
        handleClick={downloadSettingsToggler.aggressiveMode}
        isDisable={isInFireFox}
        message={message}
        testId="aggressiveMode-integration-switch"
      />
      <RichFeatureSwitch
        name={i18n('Filename Detector', 'options:integrations')}
        desc={i18n(
          'The detector can notify user when the filename is modified by other extensions.',
          'options:integrations'
        )}
        isOn={warningSettings.ignoreFilenameOverwritten}
        handleClick={warningSettingsToggler.ignoreFilenameOverwritten}
        isDisable={isInFireFox}
        message={message}
        testId="filenameDetector-integration-switch"
      />
    </VStack>
  )
}

export default IntegrationOptions
