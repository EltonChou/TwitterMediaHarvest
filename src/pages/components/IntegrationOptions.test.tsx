/**
 * @jest-environment jsdom
 */
import { MockDownloadSettingsRepository } from '#mocks/repositories/downloadSettings'
import { queryByTestId, render, waitFor } from '@testing-library/react'
import React from 'react'
import IntegrationOptions from './IntegrationOptions'

describe('unit test for IntegrationOptions component', () => {
  const downloadSettingsRepo = new MockDownloadSettingsRepository()

  afterAll(() => (process.env.TARGET = 'test'))

  it.each(['chrome', 'edge', 'firefox'])('match %s snapshot', target => {
    process.env.TARGET = target

    const { container, unmount } = render(
      <IntegrationOptions downloadSettingsRepo={downloadSettingsRepo} />
    )

    const aria2ExtLink = queryByTestId(container, 'aria2-ext-link')

    waitFor(() => {
      expect(container).toMatchSnapshot(target)
      target !== 'firefox'
        ? expect(aria2ExtLink).toBeInTheDocument()
        : expect(aria2ExtLink).not.toBeInTheDocument()
    })

    unmount()
  })
})
