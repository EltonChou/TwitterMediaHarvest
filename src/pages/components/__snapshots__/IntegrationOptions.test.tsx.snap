// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit test for IntegrationOptions component match chrome snapshot: chrome 1`] = `
<div>
  <div
    class="chakra-stack css-tl3ftk"
  >
    <div
      class="chakra-form-control css-0"
      data-testid="dispatchToAria2-feature-switch"
      label="Translated<options_integrations_dispatchToAria2>"
      role="group"
    >
      <label
        class="chakra-form__label css-28a6i0"
        data-testid="feature-switch-label"
        for=":r0:"
        id="field-:r1:-label"
      >
        <div
          class="chakra-stack css-mz2o5n"
          style="transition: background 300ms;"
        >
          <div
            class="chakra-stack css-1bef4uc"
          >
            <p
              class="chakra-text css-1tbqbam"
            >
              Translated&lt;options_integrations_dispatchToAria2&gt;
            </p>
            <p
              class="chakra-text css-qljqz"
            >
              Translated&lt;options_integrations_dispatchToAria2_desc&gt;
              <a
                class="chakra-link css-p03q1r"
                href="https://chrome.google.com/webstore/detail/mpkodccbngfoacfalldjimigbofkhgjn"
                target="_blank"
              >
                Aria2-Explorer
              </a>
            </p>
             
          </div>
          <div
            class="css-fzqoy4"
          >
            <label
              class="chakra-switch css-16pgy8f"
              data-testid="feature-switch"
            >
              <input
                aria-disabled="false"
                aria-invalid="false"
                class="chakra-switch__input"
                id=":r0:"
                style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-switch__track css-14qxnv8"
              >
                <span
                  class="chakra-switch__thumb css-0"
                />
              </span>
            </label>
          </div>
        </div>
      </label>
    </div>
    <div
      class="chakra-form-control css-0"
      data-testid="aggressiveMode-feature-switch"
      label="Translated<options_integrations_aggressiveMode>"
      role="group"
    >
      <label
        class="chakra-form__label css-28a6i0"
        data-testid="feature-switch-label"
        for=":r2:"
        id="field-:r3:-label"
      >
        <div
          class="chakra-stack css-mz2o5n"
          style="transition: background 300ms;"
        >
          <div
            class="chakra-stack css-1bef4uc"
          >
            <p
              class="chakra-text css-1tbqbam"
            >
              Translated&lt;options_integrations_aggressiveMode&gt;
            </p>
            <p
              class="chakra-text css-qljqz"
            >
              Translated&lt;options_integrations_aggressiveMode_desc&gt;
            </p>
             
          </div>
          <div
            class="css-fzqoy4"
          >
            <label
              class="chakra-switch css-16pgy8f"
              data-testid="feature-switch"
            >
              <input
                aria-disabled="false"
                aria-invalid="false"
                class="chakra-switch__input"
                id=":r2:"
                style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-switch__track css-14qxnv8"
              >
                <span
                  class="chakra-switch__thumb css-0"
                />
              </span>
            </label>
          </div>
        </div>
      </label>
    </div>
  </div>
</div>
`;

exports[`unit test for IntegrationOptions component match edge snapshot: edge 1`] = `
<div>
  <div
    class="chakra-stack css-tl3ftk"
  >
    <div
      class="chakra-form-control css-0"
      data-testid="dispatchToAria2-feature-switch"
      label="Translated<options_integrations_dispatchToAria2>"
      role="group"
    >
      <label
        class="chakra-form__label css-28a6i0"
        data-testid="feature-switch-label"
        for=":r4:"
        id="field-:r5:-label"
      >
        <div
          class="chakra-stack css-mz2o5n"
          style="transition: background 300ms;"
        >
          <div
            class="chakra-stack css-1bef4uc"
          >
            <p
              class="chakra-text css-1tbqbam"
            >
              Translated&lt;options_integrations_dispatchToAria2&gt;
            </p>
            <p
              class="chakra-text css-qljqz"
            >
              Translated&lt;options_integrations_dispatchToAria2_desc&gt;
              <a
                class="chakra-link css-p03q1r"
                href="https://chrome.google.com/webstore/detail/mpkodccbngfoacfalldjimigbofkhgjn"
                target="_blank"
              >
                Aria2-Explorer
              </a>
            </p>
             
          </div>
          <div
            class="css-fzqoy4"
          >
            <label
              class="chakra-switch css-16pgy8f"
              data-testid="feature-switch"
            >
              <input
                aria-disabled="false"
                aria-invalid="false"
                class="chakra-switch__input"
                id=":r4:"
                style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-switch__track css-14qxnv8"
              >
                <span
                  class="chakra-switch__thumb css-0"
                />
              </span>
            </label>
          </div>
        </div>
      </label>
    </div>
    <div
      class="chakra-form-control css-0"
      data-testid="aggressiveMode-feature-switch"
      label="Translated<options_integrations_aggressiveMode>"
      role="group"
    >
      <label
        class="chakra-form__label css-28a6i0"
        data-testid="feature-switch-label"
        for=":r6:"
        id="field-:r7:-label"
      >
        <div
          class="chakra-stack css-mz2o5n"
          style="transition: background 300ms;"
        >
          <div
            class="chakra-stack css-1bef4uc"
          >
            <p
              class="chakra-text css-1tbqbam"
            >
              Translated&lt;options_integrations_aggressiveMode&gt;
            </p>
            <p
              class="chakra-text css-qljqz"
            >
              Translated&lt;options_integrations_aggressiveMode_desc&gt;
            </p>
             
          </div>
          <div
            class="css-fzqoy4"
          >
            <label
              class="chakra-switch css-16pgy8f"
              data-testid="feature-switch"
            >
              <input
                aria-disabled="false"
                aria-invalid="false"
                class="chakra-switch__input"
                id=":r6:"
                style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-switch__track css-14qxnv8"
              >
                <span
                  class="chakra-switch__thumb css-0"
                />
              </span>
            </label>
          </div>
        </div>
      </label>
    </div>
  </div>
</div>
`;

exports[`unit test for IntegrationOptions component match firefox snapshot: firefox 1`] = `
<div>
  <div
    class="chakra-stack css-tl3ftk"
  >
    <div
      class="chakra-form-control css-0"
      data-disabled=""
      data-testid="dispatchToAria2-feature-switch"
      label="Translated<options_integrations_dispatchToAria2>"
      role="group"
    >
      <label
        class="chakra-form__label css-qf9lcl"
        data-disabled=""
        data-testid="feature-switch-label"
        for=":r8:"
        id="field-:r9:-label"
      >
        <div
          class="chakra-stack css-mz2o5n"
          style="transition: background 300ms;"
        >
          <div
            class="chakra-stack css-1bef4uc"
          >
            <p
              class="chakra-text css-1tbqbam"
            >
              Translated&lt;options_integrations_dispatchToAria2&gt;
            </p>
            <p
              class="chakra-text css-qljqz"
            >
              Translated&lt;options_integrations_dispatchToAria2_desc&gt;
               Aria2-Explorer 
            </p>
            <div
              class="chakra-offset-slide"
              style="opacity: 0; transform: translateX(0px) translateY(10px) translateZ(0);"
            >
              <div
                class="chakra-stack css-ng2lei"
              >
                <svg
                  class="chakra-icon chakra-icon css-raeoph"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
                    fill="currentColor"
                  />
                </svg>
                <div
                  class="chakra-form__helper-text css-s0rnn6"
                  data-testid="feature-switch-helper-text"
                  id="field-:r9:-helptext"
                >
                  Translated&lt;options_integrations_notCompatibble&gt;
                </div>
              </div>
            </div>
          </div>
          <div
            class="css-fzqoy4"
          >
            <label
              class="chakra-switch css-16pgy8f"
              data-disabled=""
              data-testid="feature-switch"
            >
              <input
                aria-describedby="field-:r9:-helptext"
                aria-disabled="true"
                aria-invalid="false"
                class="chakra-switch__input"
                disabled=""
                id=":r8:"
                style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-switch__track css-14qxnv8"
                data-disabled=""
              >
                <span
                  class="chakra-switch__thumb css-0"
                  data-disabled=""
                />
              </span>
            </label>
          </div>
        </div>
      </label>
    </div>
    <div
      class="chakra-form-control css-0"
      data-disabled=""
      data-testid="aggressiveMode-feature-switch"
      label="Translated<options_integrations_aggressiveMode>"
      role="group"
    >
      <label
        class="chakra-form__label css-qf9lcl"
        data-disabled=""
        data-testid="feature-switch-label"
        for=":rb:"
        id="field-:rc:-label"
      >
        <div
          class="chakra-stack css-mz2o5n"
          style="transition: background 300ms;"
        >
          <div
            class="chakra-stack css-1bef4uc"
          >
            <p
              class="chakra-text css-1tbqbam"
            >
              Translated&lt;options_integrations_aggressiveMode&gt;
            </p>
            <p
              class="chakra-text css-qljqz"
            >
              Translated&lt;options_integrations_aggressiveMode_desc&gt;
            </p>
            <div
              class="chakra-offset-slide"
              style="opacity: 0; transform: translateX(0px) translateY(10px) translateZ(0);"
            >
              <div
                class="chakra-stack css-ng2lei"
              >
                <svg
                  class="chakra-icon chakra-icon css-raeoph"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
                    fill="currentColor"
                  />
                </svg>
                <div
                  class="chakra-form__helper-text css-s0rnn6"
                  data-testid="feature-switch-helper-text"
                  id="field-:rc:-helptext"
                >
                  Translated&lt;options_integrations_notCompatibble&gt;
                </div>
              </div>
            </div>
          </div>
          <div
            class="css-fzqoy4"
          >
            <label
              class="chakra-switch css-16pgy8f"
              data-disabled=""
              data-testid="feature-switch"
            >
              <input
                aria-describedby="field-:rc:-helptext"
                aria-disabled="true"
                aria-invalid="false"
                class="chakra-switch__input"
                disabled=""
                id=":rb:"
                style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-switch__track css-14qxnv8"
                data-disabled=""
              >
                <span
                  class="chakra-switch__thumb css-0"
                  data-disabled=""
                />
              </span>
            </label>
          </div>
        </div>
      </label>
    </div>
  </div>
</div>
`;
